File: search_api.js
#!/usr/bin/env node
/**
 * PGroonga全文検索API実装例（Express + PostgreSQL）
 */
const express = require('express');
const { Pool } = require('pg');

const app = express();
const pool = new Pool({
  host: 'localhost',
  port: 5432,
  user: 'postgres',
  password: 'password',
  database: 'postgres'
});

app.use(express.json());

/**
 * 基本的なキーワード検索API
 * タイトルまたはコンテンツに対して全文検索を実行
 */
app.get('/search', async (req, res) => {
  const { keyword, min_views, limit = 10 } = req.query;

  if (!keyword) {
    return res.status(400).json({ error: 'keyword is required' });
  }

  let query = `
    SELECT id, title, author, view_count, published_at
    FROM blog_posts
    WHERE (title &@~ $1 OR content &@~ $1)
  `;

  const params = [keyword];
  let paramIndex = 2;

  if (min_views) {
    query += ` AND view_count >= $${paramIndex}`;
    params.push(parseInt(min_views));
    paramIndex++;
  }

  query += ` ORDER BY view_count DESC LIMIT $${paramIndex}`;
  params.push(parseInt(limit));

  try {
    const result = await pool.query(query, params);
    res.json(result.rows);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

/**
 * シーク法を使った高速ページネーション
 * 大量データでも一定の高速性を維持
 */
app.get('/search/paginated', async (req, res) => {
  const { keyword, last_published_at, last_id, limit = 10 } = req.query;

  if (!keyword) {
    return res.status(400).json({ error: 'keyword is required' });
  }

  // 1件多く取得してhas_nextを判定
  const fetchLimit = parseInt(limit) + 1;

  let query;
  let params;

  if (last_published_at && last_id) {
    // 2ページ目以降
    query = `
      SELECT id, title, author, view_count, published_at
      FROM blog_posts
      WHERE (title &@~ $1 OR content &@~ $1)
        AND (published_at, id) < ($2, $3)
      ORDER BY published_at DESC, id DESC
      LIMIT $4
    `;
    params = [keyword, last_published_at, parseInt(last_id), fetchLimit];
  } else {
    // 1ページ目
    query = `
      SELECT id, title, author, view_count, published_at
      FROM blog_posts
      WHERE (title &@~ $1 OR content &@~ $1)
      ORDER BY published_at DESC, id DESC
      LIMIT $2
    `;
    params = [keyword, fetchLimit];
  }

  try {
    const result = await pool.query(query, params);
    
    // has_nextフラグを判定
    const hasNext = result.rows.length > parseInt(limit);
    const actualResults = result.rows.slice(0, parseInt(limit));

    // 最終レコードの情報を取得
    let responseLastId = null;
    let responseLastPublishedAt = null;
    if (actualResults.length > 0) {
      const lastRecord = actualResults[actualResults.length - 1];
      responseLastId = lastRecord.id;
      responseLastPublishedAt = lastRecord.published_at;
    }

    res.json({
      results: actualResults,
      has_next: hasNext,
      last_id: responseLastId,
      last_published_at: responseLastPublishedAt
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

/**
 * タグ検索と全文検索の組み合わせ
 * 配列型のGINインデックスとPGroongaを併用
 */
app.get('/search/by-tags', async (req, res) => {
  const { tag, keyword, limit = 10 } = req.query;

  if (!tag) {
    return res.status(400).json({ error: 'tag is required' });
  }

  let query;
  let params;

  if (keyword) {
    query = `
      SELECT id, title, tags, view_count
      FROM blog_posts
      WHERE $1 = ANY(tags)
        AND (title &@~ $2 OR content &@~ $2)
      ORDER BY view_count DESC
      LIMIT $3
    `;
    params = [tag, keyword, parseInt(limit)];
  } else {
    query = `
      SELECT id, title, tags, view_count
      FROM blog_posts
      WHERE $1 = ANY(tags)
      ORDER BY view_count DESC
      LIMIT $2
    `;
    params = [tag, parseInt(limit)];
  }

  try {
    const result = await pool.query(query, params);
    res.json(result.rows);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

/**
 * 検索候補（サジェスト）機能
 * 前方一致検索でオートコンプリートを実現
 */
app.get('/suggest', async (req, res) => {
  const { prefix, limit = 5 } = req.query;

  if (!prefix) {
    return res.status(400).json({ error: 'prefix is required' });
  }

  const query = `
    SELECT DISTINCT title
    FROM blog_posts
    WHERE title &^ $1
    LIMIT $2
  `;

  try {
    const result = await pool.query(query, [prefix, parseInt(limit)]);
    res.json(result.rows.map(row => row.title));
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

File: search_api.py
#!/usr/bin/env python3
"""
PGroonga全文検索API実装例（FastAPI）
"""
from fastapi import FastAPI, Query, HTTPException
from typing import List, Optional
from datetime import datetime
import asyncpg
from pydantic import BaseModel

app = FastAPI(title="PGroonga Search API")


class SearchResult(BaseModel):
    id: int
    title: str
    author: str
    view_count: int
    published_at: datetime


class PaginatedSearchResult(BaseModel):
    results: List[SearchResult]
    has_next: bool
    last_id: Optional[int] = None
    last_published_at: Optional[datetime] = None


async def get_db_connection():
    """データベース接続を取得"""
    return await asyncpg.connect(
        host='localhost',
        port=5432,
        user='postgres',
        password='password',
        database='postgres'
    )


@app.get("/search", response_model=List[SearchResult])
async def search_posts(
    keyword: str = Query(..., min_length=1, description="検索キーワード"),
    min_views: Optional[int] = Query(None, ge=0, description="最小閲覧数"),
    limit: int = Query(10, le=100, description="取得件数")
):
    """
    基本的なキーワード検索API
    タイトルまたはコンテンツに対して全文検索を実行
    """
    conn = await get_db_connection()

    query = """
        SELECT id, title, author, view_count, published_at
        FROM blog_posts
        WHERE (title &@~ $1 OR content &@~ $1)
    """
    params = [keyword]

    if min_views is not None:
        query += " AND view_count >= $2"
        params.append(min_views)
        query += " ORDER BY view_count DESC LIMIT $3"
        params.append(limit)
    else:
        query += " ORDER BY view_count DESC LIMIT $2"
        params.append(limit)

    try:
        results = await conn.fetch(query, *params)
        return [dict(row) for row in results]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        await conn.close()


@app.get("/search/paginated", response_model=PaginatedSearchResult)
async def search_posts_paginated(
    keyword: str = Query(..., min_length=1, description="検索キーワード"),
    last_published_at: Optional[str] = Query(None, description="前ページ最終レコードのpublished_at"),
    last_id: Optional[int] = Query(None, description="前ページ最終レコードのid"),
    limit: int = Query(10, le=100, description="取得件数")
):
    """
    シーク法を使った高速ページネーション
    大量データでも一定の高速性を維持
    """
    conn = await get_db_connection()

    # 1件多く取得してhas_nextを判定
    fetch_limit = limit + 1

    if last_published_at and last_id:
        # 2ページ目以降
        query = """
            SELECT id, title, author, view_count, published_at
            FROM blog_posts
            WHERE (title &@~ $1 OR content &@~ $1)
              AND (published_at, id) < ($2, $3)
            ORDER BY published_at DESC, id DESC
            LIMIT $4
        """
        params = [keyword, last_published_at, last_id, fetch_limit]
    else:
        # 1ページ目
        query = """
            SELECT id, title, author, view_count, published_at
            FROM blog_posts
            WHERE (title &@~ $1 OR content &@~ $1)
            ORDER BY published_at DESC, id DESC
            LIMIT $2
        """
        params = [keyword, fetch_limit]

    try:
        results = await conn.fetch(query, *params)
        
        # has_nextフラグを判定
        has_next = len(results) > limit
        actual_results = results[:limit]

        # 最終レコードの情報を取得
        response_last_id = None
        response_last_published_at = None
        if actual_results:
            last_record = actual_results[-1]
            response_last_id = last_record['id']
            response_last_published_at = last_record['published_at']

        return {
            "results": [dict(row) for row in actual_results],
            "has_next": has_next,
            "last_id": response_last_id,
            "last_published_at": response_last_published_at
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        await conn.close()


@app.get("/search/by-tags")
async def search_by_tags(
    tag: str = Query(..., description="検索対象のタグ"),
    keyword: Optional[str] = Query(None, description="追加の全文検索キーワード"),
    limit: int = Query(10, le=100, description="取得件数")
):
    """
    タグ検索と全文検索の組み合わせ
    配列型のGINインデックスとPGroongaを併用
    """
    conn = await get_db_connection()

    if keyword:
        query = """
            SELECT id, title, tags, view_count
            FROM blog_posts
            WHERE $1 = ANY(tags)
              AND (title &@~ $2 OR content &@~ $2)
            ORDER BY view_count DESC
            LIMIT $3
        """
        params = [tag, keyword, limit]
    else:
        query = """
            SELECT id, title, tags, view_count
            FROM blog_posts
            WHERE $1 = ANY(tags)
            ORDER BY view_count DESC
            LIMIT $2
        """
        params = [tag, limit]

    try:
        results = await conn.fetch(query, *params)
        return [dict(row) for row in results]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        await conn.close()


@app.get("/suggest")
async def suggest_titles(
    prefix: str = Query(..., min_length=1, description="検索プレフィックス"),
    limit: int = Query(5, le=20, description="候補数")
):
    """
    検索候補（サジェスト）機能
    前方一致検索でオートコンプリートを実現
    """
    conn = await get_db_connection()

    query = """
        SELECT DISTINCT title
        FROM blog_posts
        WHERE title &^ $1
        LIMIT $2
    """

    try:
        results = await conn.fetch(query, prefix, limit)
        return [row['title'] for row in results]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        await conn.close()


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)

